src/soc/amd/common/boot/src/lib.rs

```diff
@@ -0,0 +1,222 @@
+            val = 0;
+        }
+    }
+}
+
+fn mem(w: &mut impl core::fmt::Write, a: Vec<u8, U16>) -> () {
+    let mut vals: Vec<u64, U8> = Vec::new();
+    hex(&a, &mut vals);
+
+    // I wish I knew rust. This code is shit.
+    for a in vals.iter() {
+        let m = peek(*a);
+        write!(w, "{:x?}: {:x?}\r\n", *a, m).unwrap();
+    }
+}
+
+fn ind(w: &mut impl core::fmt::Write, a: Vec<u8, U16>) -> () {
+    let mut vals: Vec<u64, U8> = Vec::new();
+    hex(&a, &mut vals);
+
+    // I wish I knew rust. This code is shit.
+    for a in vals.iter() {
+        let m = unsafe { inl(*a as u16) };
+        write!(w, "{:x?}: {:x?}\r\n", *a, m).unwrap();
+    }
+}
+
+fn out(w: &mut impl core::fmt::Write, a: Vec<u8, U16>) -> () {
+    let mut vals: Vec<u64, U8> = Vec::new();
+    hex(&a, &mut vals);
+
+    // I wish I knew rust. This code is shit.
+    for i in 0..vals.len() / 2 {
+        let a = vals[i * 2] as u16;
+        let v = vals[i * 2 + 1] as u32;
+        unsafe {
+            outl(a, v);
+        };
+        write!(w, "{:x?}: {:x?}\r\n", a, v).unwrap();
+    }
+}
+
+fn memb(w: &mut impl core::fmt::Write, a: Vec<u8, U16>) -> () {
+    let mut vals: Vec<u64, U8> = Vec::new();
+    hex(&a, &mut vals);
+    write!(w, "dump bytes: {:x?}\r\n", vals).expect("Failed to write.");
+    for a in vals.iter() {
+        for i in 0..16 {
+            let m = peekb(*a + i);
+            write!(w, "{:x?}: {:x?}\r\n", *a + i, m).unwrap();
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn _asdebug(w: &mut impl core::fmt::Write, a: u64) -> () {
+    write!(w, "here we are in asdebug\r\n").unwrap();
+    write!(w, "stack is {:x?}\r\n", a).unwrap();
+    consdebug(w);
+    write!(w, "back to hell\r\n").unwrap();
+}
+
+fn consdebug(w: &mut impl core::fmt::Write) -> () {
+    let mut done: bool = false;
+    let newline: [u8; 2] = [10, 13];
+    while done == false {
+        let mut io = IOPort {};
+        let uart0 = &mut I8250::new(0x3f8, 0, &mut io);
+        let mut line: Vec<u8, U16> = Vec::new();
+        loop {
+            let mut c: [u8; 1] = [12; 1];
+            uart0.pread(&mut c, 1).unwrap();
+            uart0.pwrite(&c, 1).unwrap();
+            line.push(c[0]).unwrap();
+            if c[0] == 13 || c[0] == 10 || c[0] == 4 {
+                uart0.pwrite(&newline, 2).unwrap();
+                break;
+            }
+            if line.len() > 15 {
+                break;
+            }
+        }
+        match line[0] {
+            0 | 4 => {
+                done = true;
+            }
+            b'm' => {
+                mem(w, line);
+            }
+            b'i' => {
+                ind(w, line);
+            }
+            b'o' => {
+                out(w, line);
+            }
+            b'h' => {
+                memb(w, line);
+            }
+            _ => {}
+        }
+    }
+}
+//global_asm!(include_str!("init.S"));
+
+pub fn boot(w: &mut impl core::fmt::Write, fdt_address: usize) {
```

Comment:
The more we separate this stuff ... the more I feel like passing the writer in as a paramater might be better? I don't know however.

Reply:
Yes. I have to agree: Passing the writer explicitly every time isn't that bad.

Let's do that for now. We could always change over later anyway.